tests/test_integration.py

```diff
@@ -0,0 +1,137 @@
+import sys
+import socket
+from threading import Thread
+import json
+
+import stripe
+import pytest
+
+if sys.version_info[0] < 3:
+    from BaseHTTPServer import BaseHTTPRequestHandler, HTTPServer
+else:
+    from http.server import BaseHTTPRequestHandler, HTTPServer
+
+
+def get_free_port():
+    s = socket.socket(socket.AF_INET, type=socket.SOCK_STREAM)
+    s.bind(("localhost", 0))
+    address, port = s.getsockname()
+    s.close()
+    return port
+
+
+class TestIntegration(object):
+    @pytest.fixture(autouse=True)
+    def close_mock_server(self):
+        yield
+        if self.mock_server:
+            self.mock_server.shutdown()
+            self.mock_server_thread.join()
+
+    @pytest.fixture(autouse=True)
+    def setup_stripe(self):
+        orig_attrs = {
+            "api_base": stripe.api_base,
+            "api_key": stripe.api_key,
+            "default_http_client": stripe.default_http_client,
+            "log": stripe.log,
+            "proxy": stripe.proxy,
+        }
+        stripe.api_base = "http://localhost:12111"  # stripe-mock
+        stripe.api_key = "sk_test_123"
+        stripe.default_http_client = None
+        stripe.log = "warning"
+        stripe.proxy = None
+        yield
+        stripe.api_base = orig_attrs["api_base"]
+        stripe.api_key = orig_attrs["api_key"]
+        stripe.default_http_client = orig_attrs["default_http_client"]
+        stripe.log = orig_attrs["log"]
+        stripe.proxy = orig_attrs["proxy"]
+
+    def setup_mock_server(self, handler):
+        # Configure mock server.
+        self.mock_server_port = get_free_port()
+        self.mock_server = HTTPServer(
+            ("localhost", self.mock_server_port), handler
+        )
```

Comment:
Yeesh, I was surprised to find that Python's docs for this are astoundingly awful, but passing `0` here should have the server select a port, which allows you to remove `get_free_port` and a possible race condition.

Here's an example with `ThreadedTCPServer`:

``` python
if __name__ == "__main__":
    # Port 0 means to select an arbitrary unused port
    HOST, PORT = "localhost", 0

    server = ThreadedTCPServer((HOST, PORT), ThreadedTCPRequestHandler)
    ip, port = server.server_address

    # Start a thread with the server -- that thread will then start one
    # more thread for each request
    server_thread = threading.Thread(target=server.serve_forever)
    # Exit the server thread when the main thread terminates
    server_thread.daemon = True
    server_thread.start()
    print "Server loop running in thread:", server_thread.name

    client(ip, port, "Hello World 1")
    client(ip, port, "Hello World 2")
    client(ip, port, "Hello World 3")

    server.shutdown()
    server.server_close()
```